File: executor/src/main/java/com/codeloom/executor/engine/DockerImageManager.java
package com.codeloom.executor.engine;

import com.github.dockerjava.api.DockerClient;
import com.github.dockerjava.api.exception.NotFoundException;
import java.util.concurrent.TimeUnit;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class DockerImageManager {
    private final DockerClient docker;

    public void pullImageIfAbsent(String image) {
        pullImageIfAbsent(image, 120);
    }

    public void pullImageIfAbsent(String image, long timeout) {
        try {
            docker.inspectImageCmd(image).exec();
        } catch (NotFoundException e) {
            log.info("Image {} not found. Pulling...", image);
            try {
                docker.pullImageCmd(image).start().awaitCompletion(timeout, TimeUnit.SECONDS);
            } catch (InterruptedException x) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException(x);
            }
            log.info("Image {} is pulled!", image);
        }
    }
}

File: executor/src/main/kotlin/com/codeloom/executor/engine/DockerImageManager.kt
package com.codeloom.executor.engine

import com.github.dockerjava.api.DockerClient
import com.github.dockerjava.api.exception.NotFoundException
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit

@Component
class DockerImageManager(
    private val dockerClient: DockerClient,
) {
    companion object {
        private val logger = LoggerFactory.getLogger(javaClass)
    }

    fun pullImageIfAbsent(
        image: String,
        timeoutSeconds: Long = 120,
    ) {
        try {
            dockerClient.inspectImageCmd(image).exec()
        } catch (_: NotFoundException) {
            logger.info("Image $image not found. Pulling...")
            dockerClient.pullImageCmd(image)
                .start()
                .awaitCompletion(timeoutSeconds, TimeUnit.SECONDS)
            logger.info("Image $image is pulled!")
        }
    }
}
